Global UI Responsive

isacoder opened the pull request:
### Requirements

[Bugherd#42](https://www.bugherd.com/projects/216308/kanban)
[Main Menu] Make responsive
Create a master component for the global UI pattern. As part of this it should fix the behaviour of how the main menu 

**Specs**
https://future-standard.slack.com/files/T0ALXERMZ/F021XG9NN8P?origin_team=T0ALXERMZ

### Implementation


- New Navigation template component is based in the traffic counter saas as mentioned.
This component eventually will handle the switch between the desktop and the mobile version.
Honestly if you have a better suggestion for the name of this template please share it 😄.


- I created two hooks **useMedia** and **useBreakpoints** for the JS management of the media queries but I added the devices values in the styled components theme, this way we can consistently use management of the same breakpoints with CSS and JS. All these based on the requirements provided. Does these hooks should be available for the UI library or should just remain just for internal code usage?.

- Added a useReducer in a useMenu hook to manage the  open-hide management state of the menu.
I didn't included the focus management on this hook because I saw a comment of a missing refocus of the context so I though it would be better to refactor this part together with the missing functionality. I recommend we create a separate task for that.

- In order to handle the menu over the content I made the Content position absolute and had to remove the "Flex-basis" in the MainMenu with a "Width" because it was not re-sizing correctly. I'm not very familiar with the usage of Flex-basis and I'm a bit worried the implementation is missing a requirement that Flex basis was in charged of. Please let me know if there is something missing. 

- The Layout page had Content variable already in place, I purposely use the Content2 naming to ask if this should be part of the layouts or better to keep the values in the template. 

### Screenshoots

Made a example page and the storybook .. but probably we should only keep the storybook 😅

Examples page 

<img width="1918" alt="Screen Shot 2021-05-26 at 10 45 09" src="https://user-images.githubusercontent.com/10409078/119590662-27b72400-be10-11eb-9668-9d42762149dd.png">

Storybook

<img width="1481" alt="Screen Shot 2021-05-26 at 11 18 42" src="https://user-images.githubusercontent.com/10409078/119592862-3d2e4d00-be14-11eb-8adc-7762bd48f349.png">

atomworks reviewed (changes requested):
I've run through all the code and it's looking solid! Got a few comments (as always) to try and refine some things in code and in usability.

Biggest issue for me right now, we have the visual glitching issue when menus and submenus open and close for the first times. 😭  Is this just a caused by us now wrapping the content? How hard do you think it will be to eliminate this?

Also, how hard would it be to add some different dummy pages to the example (or even better, the story)? They don't need to be anything clever (page title component & different lipsum paragraphs maybe), just different enough to demonstrate the interactions working and how to use the wrapper itself with content. I have no real experience with React Router so if this isn't a job that's pretty quick and easy, let's not worry at the moment.

I have some other observations but I don't think anything worth sharing now. I think some of those might even be their own future tasks. 

Thanks for you hard work on this!

isacoder commented on src/themes/common.tsx:
```diff
@@ -82,4 +82,23 @@ export const animation = {
+// };
+
+
+export const deviceSize = {
+  mobileS: `(max-width: 640px)`,
+  mobileM: `(min-width: 321px)`,
+  tabletP: `(min-width: 768px)`,
+  desktopS: `(min-width: 1024px)`,
+  desktop: `(min-width: 1280px)`,
+  desktopL: `(min-width: 1400px)`
 };
```
First let me apologize for the typo I have in the first media query it should be 319
```
mobileS: `(max-width: 319px)`,
```

My biggest concern here was to get a way to have both values for JS and CSS.

I made a search about different options and my favorite was this one
[how-to-use-media-queries-with-styled-components](https://jsramblings.com/how-to-use-media-queries-with-styled-components/#:~:text=Media%20queries%20with%20styled%20components%20work%20the%20same%20as%20in,different%20device%20sizes%2C%20continue%20below.)

There are two mobiles becase I saw there was two mobile versions in the Zepplin with some small sizes variations

MobileS
https://zpl.io/2GJ6Omr

MobileM
https://zpl.io/ag4XNnA

About the naming  small, medium, large, xlarge and xxLarge works for me, I made them mainly based on the requirements document you shared, so I'll go ahead and make this update.

isacoder commented on storybook/src/stories/Global/NavigationTemplate.stories.tsx:
```diff
@@ -0,0 +1,118 @@
+  ])
+
+  return (
+    <NavigationTemplate
```
updated!

isacoder commented on src/hooks/useBreakpoints.ts:
```diff
@@ -0,0 +1,45 @@
+| 'tabletP'
+| 'desktopS'
+| 'desktop'
+| 'desktopL';
```
updated!

isacoder commented on src/Global/atoms/Layout.tsx:
```diff
@@ -27,4 +27,13 @@ export const MainContainer = styled.div`
   flex-direction: column;
+`;
+
+export const Content2 = styled.div`
```
Updated to ContentArea

isacoder commented:
I have updated all requested, added the pages to Storybook and removed the example page.
After the review of the Glich with Josh we agreed that this is something that currently lives in the base code.
We have not seen it much because current Apps does not have a lot of options in menus.

Therefore, we do need to fix this but perhaps we can add this PR and follow up this issue in another.

atomworks reviewed (approved):
Awesome. Thanks for those changes. Looks great.

JoshLipps reviewed (approved):
LGTM
